src/sdk/pynni/nni/compression/torch/utils/mask_conflict.py

```diff
@@ -1,16 +1,187 @@
+        the traced model of the target model, is this parameter is not None,
+        we donnot use the model and dummpy_input to get the trace graph.
+    """
+    if isinstance(masks, str):
+        # if the input is the path of the mask_file
+        assert os.path.exists(masks)
+        masks = torch.load(masks)
+    # if the user uses the model and dummy_input to trace the model, we
+    # should get the traced model handly, so that, we only trace the
+    # model once, GroupMaskConflict and ChannelMaskConflict will reuse
+    # this traced model.
+    if traced is None:
+        assert model is not None and dummy_input is not None
+        with torch.onnx.set_training(model, False):
+            # We need to trace the model in this way, else it will have problems
+            traced = torch.jit.trace(model, dummy_input)
+
+    fix_group_mask = GroupMaskConflict(masks, model, dummy_input, traced)
+    masks = fix_group_mask.fix_mask()
+    fix_channel_mask = ChannelMaskConflict(masks, model, dummy_input, traced)
+    masks = fix_channel_mask.fix_mask()
+    padding_cat_mask = CatMaskPadding(masks, model, dummy_input, traced)
+    masks = padding_cat_mask.fix_mask()
+    return masks
+
+class MaskFix:
+    def __init__(self, masks, model=None, dummy_input=None, traced=None):
+        # check if the parameters are valid
+        parameter_valid = False
+        if traced is not None:
+            parameter_valid = True
+        elif (model is not None) and (dummy_input is not None):
+            parameter_valid = True
+        if not parameter_valid:
+            raise Exception('The input parameters is invalid!')
+        self.model = model
+        self.dummy_input = dummy_input
+        self.traced = traced
+        self.masks = masks
+
+    def fix_mask(self):
+        raise NotImplementedError
+
+    def export(self, path):
+        """
+        Export the masks after fixing the conflict to file.
+        """
+        torch.save(self.masks, path)
+
+class CatMaskPadding(MaskFix):
+    def __init__(self, masks, model, dummy_input=None, traced=None):
+        super(CatMaskPadding, self).__init__(masks, model, dummy_input, traced)
+
+    def fix_mask(self):
+        cat_padding_depen = CatPaddingDependency(self.model, self.dummy_input, self.traced)
+        name_to_module = {}
+        for name, module in self.model.named_modules():
+            name_to_module[name] = module
+        depen = cat_padding_depen.dependency_sets
+        for layers in depen:
+            device = None
+            count = 0
+            for layer in layers:
+                if layer in self.masks:
+                    count += 1
+                    if device is None:
+                        device = self.masks[layer]['weight'].device
+            if count == 0:
+                # no layer is pruned
+                continue
+            elif count == len(layers):
+                # all the layers have been pruned
```

Comment:
even all the layers have been pruned, is it possible their masks are still not consistent?

Reply:
got it.